src/Control/Monad/Metrics/Internal.hs

```diff
@@ -25,46 +36,114 @@ import           System.Metrics.Distribution (Distribution)
 import           System.Metrics.Gauge        (Gauge)
 import           System.Metrics.Label        (Label)
 
+type IdenticalKeyTypes m m' =
+    ( CounterKey m      ~ CounterKey m'
+    , GaugeKey m        ~ GaugeKey m'
+    , DistributionKey m ~ DistributionKey m'
+    , LabelKey m        ~ LabelKey m'
+    )
+
+type ValidMetricKeys m =
+    ( MetricKey (CounterKey m)
+    , MetricKey (GaugeKey m)
+    , MetricKey (DistributionKey m)
+    , MetricKey (LabelKey m)
+    )
+
+type CounterKeyFor      m key = (MetricKey key, key ~ CounterKey m)
+type GaugeKeyFor        m key = (MetricKey key, key ~ GaugeKey m)
+type DistributionKeyFor m key = (MetricKey key, key ~ DistributionKey m)
+type LabelKeyFor        m key = (MetricKey key, key ~ LabelKey m)
+
+-- | We can use arbitrary key types for our 'MonadMetrics'es, but to do so we
+-- need to a way to look the types up. Sometimes we want this association even
+-- for a type that can't be a 'MonadMetrics', like 'IO'.
+--
+-- 'Meterable' provides that capability. It has a 'MonadTrans' instance, so
+-- transformer stacks will use the key types for the monad they build upon.
+class Monad m => Meterable m where
+    type CounterKey      m :: *
+    type GaugeKey        m :: *
+    type DistributionKey m :: *
+    type LabelKey        m :: *
+
+instance Meterable IO where
+    type CounterKey IO      = Text
+    type GaugeKey IO        = Text
+    type DistributionKey IO = Text
+    type LabelKey IO        = Text
+
+instance {-# OVERLAPPABLE #-} (Meterable m, MonadTrans t, Monad (t m)) => Meterable (t m) where
+    type CounterKey (t m)      = CounterKey m
+    type GaugeKey (t m)        = GaugeKey m
+    type DistributionKey (t m) = DistributionKey m
+    type LabelKey (t m)        = LabelKey m
+
+-- | A type can be an instance of 'MonadMetrics' if it can provide a 'Metrics'
+-- somehow, and has associated key types. Commonly, this will be implemented as a 'ReaderT' where some
+-- field in the environment is the 'Metrics' data.
+--
+-- * /Since v0.1.0.0/
+class Meterable m => MonadMetrics m where
+    getMetrics      :: m (KeyedMetrics (CounterKey m) (GaugeKey m) (DistributionKey m) (LabelKey m))
+
+instance {-# OVERLAPPABLE #-} (MonadMetrics m, MonadTrans t, Monad (t m), Meterable (t m)) => MonadMetrics (t m) where
+    getMetrics = lift getMetrics
+
+instance (Monad m, CounterKeyFor m ck, GaugeKeyFor m gk, DistributionKeyFor m dk, LabelKeyFor m lk, Meterable m) => MonadMetrics (ReaderT (KeyedMetrics ck gk dk lk) m) where
+    getMetrics = ask
```

Comment:
This instance looks like it defers to the `m` for the `Meterable` stuff, so `ReaderT Metrics IO` would have the `IO` instance for `Meterable`, and `Text` for all keys. So `ReaderT (KeyedMetrics a b c d) IO` would only work if they're all `Text`. I think we'd want to have an instance `Meterable (ReaderT (Metrics a b c d) m)` that defines it in terms of the `Reader`, and let clients implement their own stacks however they want.

Reply:
If we do that, then getting the type right for `run'` is going to be nasty. This instance basically exists in support of `run'`.